pgrx/src/callconv.rs

```diff
@@ -283,3 +283,182 @@ where
+        default: DefaultValue,
+    ) -> PgBox<ReturnType> {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe {
+            let fcinfo = PgBox::from_pg(self.0);
+            let mut flinfo = PgBox::from_pg(fcinfo.flinfo);
+            if flinfo.fn_extra.is_null() {
+                flinfo.fn_extra = PgMemoryContexts::For(flinfo.fn_mcxt)
+                    .leak_and_drop_on_delete(default())
+                    as crate::void_mut_ptr;
+            }
+
+            PgBox::from_pg(flinfo.fn_extra as *mut ReturnType)
+        }
+    }
+
+    #[inline]
+    pub fn srf_is_first_call(&self) -> bool {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe { (*(*self.0).flinfo).fn_extra.is_null() }
+    }
+
+    /// Thin wrapper around [`pg_sys::init_MultiFuncCall`], made necessary
+    /// because this structure's FunctionCallInfo is a private field.
+    #[inline]
+    pub unsafe fn init_multi_func_call(&mut self) -> FcContext<'fcx> {
+        unsafe { FcContext::assume_valid(pg_sys::init_MultiFuncCall(self.0)) }
+    }
+
+    /// Thin wrapper around [`pg_sys::per_MultiFuncCall`], made necessary
+    /// because this structure's FunctionCallInfo is a private field.
+    #[inline]
+    pub unsafe fn per_multi_func_call(&mut self) -> FcContext<'fcx> {
+        unsafe { FcContext::assume_valid(pg_sys::per_MultiFuncCall(self.0)) }
+    }
+    /// Equivalent to "per_MultiFuncCall" with no FFI cost, and a lifetime
+    /// constraint.
+    ///
+    /// Safety: This is only equivalent to the current (as of Postgres 16.3)
+    /// implementation of [`pg_sys::per_MultiFuncCall`], and future changes
+    /// to the complexity of the calling convention may break this method.
+    pub(crate) unsafe fn deref_fcx(&mut self) -> FcContext<'fcx> {
+        unsafe { FcContext::assume_valid((*(*self.0).flinfo).fn_extra.cast()) }
+    }
+
+    #[inline]
+    pub unsafe fn srf_return_next(&mut self) {
+        unsafe {
+            let mut fncctx = self.deref_fcx();
+            (*fncctx.get_inner()).call_cntr += 1;
+            (*((*self.0).resultinfo as *mut pg_sys::ReturnSetInfo)).isDone =
+                pg_sys::ExprDoneCond_ExprMultipleResult;
+        }
+    }
+
+    #[inline]
+    pub unsafe fn srf_return_done(&mut self) {
+        unsafe {
+            let mut fncctx = self.deref_fcx();
+            pg_sys::end_MultiFuncCall(self.0, fncctx.get_inner());
+            (*((*self.0).resultinfo as *mut pg_sys::ReturnSetInfo)).isDone =
+                pg_sys::ExprDoneCond_ExprEndResult;
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct FcContext<'fcx>(
+    *mut pgrx_pg_sys::FuncCallContext,
+    std::marker::PhantomData<&'fcx mut pgrx_pg_sys::FuncCallContext>,
+);
```

Comment:
It's not clear to me that we can't just offer `FcContext` as `&mut pg_sys::FuncCallContext` as unlike the fcinfo, the callee "owns" this.

Reply:
Rewrote it as that.